src/dh/clu_dh.c

```diff
@@ -0,0 +1,393 @@
+        if (wc_DhGenerateParams(&rng, modSz, &dh) != 0) {
+            WOLFCLU_LOG(WOLFCLU_E0, "Error generating parameters");
+            ret = WOLFCLU_FATAL_ERROR;
+        }
+    }
+
+    /* print out the dh parameters */
+    if (ret == WOLFCLU_SUCCESS && !noOut) {
+        byte* outBuf = NULL;
+        byte* pem    = NULL;
+        word32 outBufSz = 0;
+        word32 pemSz    = 0;
+
+        if (wc_DhParamsToDer(&dh, NULL, &outBufSz) != LENGTH_ONLY_E) {
+            WOLFCLU_LOG(WOLFCLU_E0, "Unable to get output buffer size");
+            ret = WOLFCLU_FATAL_ERROR;
+        }
+
+        if (ret == WOLFCLU_SUCCESS) {
+            outBuf = (byte*)XMALLOC(outBufSz, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+            if (outBuf == NULL) {
+                ret = WOLFCLU_FATAL_ERROR;
+            }
+        }
+
+        if (ret == WOLFCLU_SUCCESS &&
+                wc_DhParamsToDer(&dh, outBuf, &outBufSz) <= 0) {
+            ret = WOLFCLU_FATAL_ERROR;
+        }
+
+        if (ret == WOLFCLU_SUCCESS) {
+            pemSz = wc_DerToPem(outBuf, outBufSz, NULL, 0, DH_PARAM_TYPE);
+            if (pemSz > 0) {
+                pem = (byte*)XMALLOC(pemSz, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+                if (pem == NULL) {
+                    ret = WOLFCLU_FATAL_ERROR;
+                }
+            }
+            else {
+                ret = WOLFCLU_FATAL_ERROR;
+            }
+        }
+
+        if (ret == WOLFCLU_SUCCESS) {
+            pemSz = wc_DerToPem(outBuf, outBufSz, pem, pemSz, DH_PARAM_TYPE);
+            if (pemSz <= 0) {
+                ret = WOLFCLU_FATAL_ERROR;
+            }
+        }
+
+        if (ret == WOLFCLU_SUCCESS &&
+                wolfSSL_BIO_write(bioOut, pem, pemSz) <= 0) {
+            ret = WOLFCLU_FATAL_ERROR;
+        }
+
+        if (pem != NULL)
+            XFREE(pem, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+        if (outBuf != NULL)
+            XFREE(outBuf, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+    }
+   
+     /* Check if parameters are valid */
+    if(ret == WOLFCLU_SUCCESS && check){
+        byte p[MAX_DH_BITS/8];
+        byte g[MAX_DH_BITS/8];
+        byte q[MAX_DH_Q_SIZE/8];
+        word32 p_len, g_len, q_len;
+
+        p_len = sizeof(p);
+        q_len = sizeof(q);
+        g_len = sizeof(g);
+        
+        /* Export DH parameters */
+        if (wc_DhExportParamsRaw(&dh, p, &p_len, q, &q_len, g, &g_len) != 0) {
+            WOLFCLU_LOG(WOLFCLU_E0, "Failed to export DH params");
+            ret = WOLFCLU_FATAL_ERROR;
+        }
+
+        if (wc_DhSetCheckKey(&dh, p, p_len, g, g_len, q, q_len, 0, &rng) != 0) {
+            WOLFCLU_LOG(WOLFCLU_E0, "Failed to set/check DH params");
+            ret = WOLFCLU_FATAL_ERROR;
+
+        }
+        else{
+
+        }
+        
+     }
+
+    /* print out the dh key */
+    if (ret == WOLFCLU_SUCCESS && genKey) {
+        byte priv[MAX_DH_BITS/8];
+        byte pub[MAX_DH_BITS/8]; 
+        word32 privSz   = sizeof(priv);
```

Comment:
these return values should be cast to avoid warnings on some compilers

Reply:
Fixed